Add warning header

nick-bulleid opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #7407 

## Proposed changes

- Create a new header that appears above the revision grid
- Use this to indicate when in rebase/merge/patch/biset modes
- Remove status bar

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://user-images.githubusercontent.com/19927400/69675247-1e6fbe80-1096-11ea-8877-7b193e77740f.png)

### After

- No conflicts
![rebase_header](https://user-images.githubusercontent.com/19927400/71805778-f6915480-305e-11ea-9a5a-ae2a7b1e0722.PNG)

- With conflicts
![rebase_header_conflicts](https://user-images.githubusercontent.com/19927400/71805799-0741ca80-305f-11ea-967f-0bb5a35553ed.PNG)


## Test methodology <!-- How did you ensure quality? -->

- Manual testing


## Test environment(s) <!-- Remove any that don't apply -->

- GIT 2.14.1
- Windows 10

<!-- Mention language, UI scaling, or anything else that might be relevant -->


----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

pmiossec commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+                switch (_state)
+                {
+                    case State.Bisect:
+                        action = "Bisect";
```
This string (and the others below) will need to be translated using `TranslationString`.

nick-bulleid replied:
Can I build the TranslationString as I currently am doing, or do they need to be defined in full?

pmiossec replied:
I think that a full sentence is better from a translation point of view.
You could do 2 sentences with and without conflicts with the 'string.Format()' format to replace the 'action': "{0} currently in progress with merge conflicts"

nick-bulleid replied:
Yes, that sounds perfect, thanks!

pmiossec commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+
+        private void AbortButton_Click(object sender, EventArgs e)
+        {
+            GitModuleForm form = FindForm() as GitModuleForm;
```
I must be in front of my IDE to be sure but, except if `FormProcess.ShowDialog()` works well with a null value, I prefer using an explicit cast instead of a 'as'. It's easier to debug a cast exception than a NRE.

What do you think?

nick-bulleid replied:
Sure, this is my first time in C# so happy to learn 🙂

RussKie replied:
ditto, as above

pmiossec commented on contributors.txt:
```diff
@@ -107,3 +107,4 @@ YYYY/MM/DD, github id, Full name, email
 2019/10/18, RalphJSmart, Ralph J Smart, id1508-gitextensions_github(at)yahoo.com
 2019/10/24, jvitkauskas, Julius Vitkauskas, zadintuvas(at)gmail.com
 2019/10/31, rickaas, Ricky Lindeman, saakcir(at)gmail.com
+2019/11/26, nick-bulleid, Nick Bulleid, nickedb(at)gmail.com
```
We generally prefer to create another PR with this change to be able to merge it quickly and prevent conflicts on this file. Please consider doing it...

RussKie commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -379,6 +379,16 @@ public static ArgumentString SkipRebaseCmd()
+        {
+            return new GitArgumentBuilder("merge") { "--abort" };
+        }
+
```
tests?

RussKie commented on GitUI/CommandsDialogs/FormBrowse.Designer.cs:
```diff
@@ -1967,5 +1968,7 @@ private void InitializeComponent()
         private ToolStripMenuItem commitInfoRightwardMenuItem;
         private ToolStripMenuItem tsmiTelemetryEnabled;
+        private Panel RevisionGridContainer;
+        private UserControls.WarningHeader RevisionHeader;
```
```suggestion
        private UserControls.WarningHeader RebaseProgressWarning;
```

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+
+namespace GitUI.UserControls
+{
+    public partial class WarningHeader : UserControl
```
If a maintainer browses the codebase and sees a file named `WarningHeader` the maintainer is unlikely to understand what this type is.
I suggest to rename to something like `RebaseProgressWarning` as such name communicates the control's purpose clearer.

nick-bulleid replied:
Sure, though it can display more than just rebase progress. I wasn't sure of the right word to capture all the states we display, ActionProgressHeader? RepositoryStateHeader?

pmiossec replied:
RevisionGridWarningHeader? Or RevisionGridProgressHeader ?

RussKie replied:
I wouldn't call it "header" at all. Today it is placed above the revision grid, tomorrow we may move it to another place. I wouldn't want to rename the control.

RussKie replied:
`InteractiveGitOperationProgressNotification`?

nick-bulleid replied:
I tried to keep it a bit shorter so went with InteractiveGitActionControl

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+            // get the current state of the repo
+            await TaskScheduler.Default;
+
+            if (module.IsValidGitWorkingDir())
```
Please invert to reduce nesting

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+            {
+                if (module.InTheMiddleOfBisect())
+                {
+                    await SetStateAsync(State.Bisect, false);
```
Add `return` here for fast return and remove unnecessary `else`
Applies everywhere.

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+                }
+            }
+
+            async Task SetStateAsync(State state, bool conflicted)
```
Please make it an instance method, and have it tested, exposing the method via `TestAccessor`.

nick-bulleid replied:
I'm having a little trouble with this. I wonder if you could give any guidance on how best to expose the State enum, and this async function for testing?

I presume I would then test all properties that I set such as _status, _hasConflicts, BackColour, ButtonContainer.Controls and Label.Text?

mstv replied:
Currently, it's a nested function (even a closure).
Make it an instance method, i.e. factor it out as a private member function of this class.
Then apply the `TestAccessor` pattern like e.g. at the bottom of `FormCommit.cs`.
The visibility level of the `enum` can be raised to `internal`.

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+                        break;
+                }
+
+                if (_state != State.None)
```
Please invert and return early.

Also the code after line:77 isn't changing the value of `_state`. So why are we not calling `Hide()` and exiting after line:77?

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+        private void ResolveButton_Click(object sender, EventArgs e)
+        {
+            GitModuleForm form = FindForm() as GitModuleForm;
+            form.UICommands.StartResolveConflictsDialog(this);
```
❗️ Line:131 carries a risk of `form` being `null`.
```suggestion
            form?.UICommands.StartResolveConflictsDialog(this);
```

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+
+        private void ContinueButton_Click(object sender, EventArgs e)
+        {
+            GitModuleForm form = FindForm() as GitModuleForm;
```
ditto
if `form` is `null` we must exit.

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+                case State.Patch:
+                    FormProcess.ShowDialog(form, GitCommandHelpers.AbortCmd());
+                    break;
+            }
```
ditto

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+                case State.Patch:
+                    FormProcess.ShowDialog(form, GitCommandHelpers.ResolvedCmd());
+                    break;
+            }
```
if `default` we must exit

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+
+        private void MoreButton_Click(object sender, EventArgs e)
+        {
+            GitModuleForm form = FindForm() as GitModuleForm;
```
ditto, as above

RussKie commented on GitUI/UserControls/WarningHeader.cs:
```diff
@@ -0,0 +1,193 @@
+            InitializeComponent();
+        }
+
+        public async Task RefreshAsync(GitCommands.GitModule module)
```
We mustn't pass `GitModule` as a parameter as it causes all sorts of dramas, including seriously inhibiting tests.
Instead please inherit the control from `GitModuleControl` and you will have access to the current instance of `GitModule`.

gerhardol commented:
I believe I prefer the status bar but I will not vote against changing this.

RussKie commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -960,101 +949,13 @@ void ShowRevisions()
         {
-            CheckForMergeConflicts();
-            return;
+            ThreadHelper.ThrowIfNotOnUIThread();
```
Everything on this point onwards to `ThreadHelper.JoinableTaskFactory.RunAsync(...)` had to run on UI thread. Now all of that functionality is gone from here, so this check is now redundant.

RussKie commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -960,101 +949,13 @@ void ShowRevisions()
 
-            void CheckForMergeConflicts()
+            // check if we are in the middle of an action (merge/rebase/bisect)
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
```
I think it can be shortened to
```cs
ThreadHelper.JoinableTaskFactory.RunAsync(RevisionHeader.RefreshAsync).FileAndForget();
```

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+        private readonly TranslationString _merge = new TranslationString("Merge");
+        private readonly TranslationString _patch = new TranslationString("Patch");
+
+        public enum State
```
These are git operations, not a repo states
Perhaps it should be called 
```suggestion
        public enum GitOperation
```

The rename would flow into the field name and parameter names.

nick-bulleid replied:
I went with GitAction to fit with the name of the control

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+        private State _state = State.None;
+        private bool _hasConflicts = false;
+
+        public event EventHandler BisectEvent;
```
Why is it `public`? How it is expected to be used?

nick-bulleid replied:
I couldn't work out how to open FormBisect from within the control, so this event tells FormBrowse to open it.

RussKie replied:
IIUIC, you delegate it to accommodate this:
```cs
        private void BisectClick(object sender, EventArgs e)
        {
            using (var frm = new FormBisect(RevisionGrid))
            {
                frm.ShowDialog(this);
            }

            UICommands.RepoChangedNotifier.Notify();
        }
```

Is there a reason we can't do this in this module?
```cs
        private void BisectClick(object sender, EventArgs e)
        {
            if (!(FindForm() is FormBrowse form))
            {
                return;
            }

            using (var frm = new FormBisect(form.RevisionGridControl))
            {
                frm.ShowDialog(this);
            }

            form.UICommands.RepoChangedNotifier.Notify();
        }
```

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+        }
+
+        private State _state = State.None;
+        private bool _hasConflicts = false;
```
Fields get auto-initialised to defaults, so initialisations are redundant.

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+            InitializeComplete();
+        }
+
+        public async Task RefreshAsync()
```
I don't really see any reasons for this method to be `async` - we are not doing any processing that should be on a background thread.
Switching between threads is not cheap.

All calls on `Module` instance are generally cheap - checking for presence of specific files. If for whatever reason we can't check these files (e.g. I/O operations take too long) than a user has a bigger issues to worry about than our UI not updating immediately.

nick-bulleid replied:
I was trying to avoid reintroducing the bug that was fixed by #2862. The issue there was calling InTheMiddleOfConflictedMerge on the UI thread, so an alternative would be to only use the background thread when calling that function?

nick-bulleid replied:
Actually, it seems that #2862 references the deprecated RunGitCmd function, so perhaps this is now fixed in a different way.

nick-bulleid replied:
I've removed all async functions and I couldn't repro the issue in #2862, but it might be worth someone who could see the issue originally taking a look.

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+            await SetStateAsync(State.None, false);
+        }
+
+        private async Task SetStateAsync(State state, bool conflicted)
```
As above, there is nothin in this method that requires switching threads, we should be on UI thread here, so no `async` is necessary.

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+                    _progressMessage.Text,
+                action);
+
+            Show();
```
Instead of `Show/Hide` you can use `Visible = true/false`

nick-bulleid replied:
I tend to prefer functional forms rather than accessing variables directly, but happy to change it if that's the standard here.

RussKie replied:
`Visible` is a property, so it incapsulates variables.

If you look at the code, [`Show()` calls `Visilbe = true`](https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs#L11378-L11381), and [`Hide()` calls `Visible = false`](https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs#L6214-L6217).

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+                case State.Patch:
+                    FormProcess.ShowDialog(form, GitCommandHelpers.ResolvedCmd());
+                    break;
+                default: return;
```
```suggestion
                default: 
                	return;
```

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+                case State.Patch:
+                    FormProcess.ShowDialog(form, GitCommandHelpers.AbortCmd());
+                    break;
+                default: return;
```
```suggestion
                default: 
                	return;
```

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+                {
+                    await that._interactiveGitActionControl.SetStateAsync(state, conflicts);
+                });
+            }
```
The `TestAccessor` is a proxy for private members, any smarts that may be necessary should be done in tests.
So we would have it:
```cs
internal void SetState(State state, bool conflicts) => _interactiveGitActionControl.SetState(state, conflicts);
```

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,240 @@
+                return;
+            }
+
+            bool conflicted = Module.InTheMiddleOfConflictedMerge();
```
```suggestion
            bool hasConflicts = Module.InTheMiddleOfConflictedMerge();
```

RussKie commented on UnitTests/GitUITests/UserControls/InteractiveGitActionControlTests.cs:
```diff
@@ -0,0 +1,36 @@
+
+            Assert.AreEqual(state, _accessor.State);
+            Assert.AreEqual(conflicts, _accessor.HasConflicts);
+        }
```
We need to add other tests to verify that:
- the control is shown/hidden
- buttons added or not
- a specific button is visible for a specific action

drewnoakes commented:
I like this feature! Great to get rid of the status bar at the bottom as well.

On the UI side I would vote for removing the 1px border from the notification panel, and using colour/contrast to bring attention to it. Maybe a ⚠️ icon or something like that too.

nick-bulleid commented:
> 
> 
> I like this feature! Great to get rid of the status bar at the bottom as well.
> 
> On the UI side I would vote for removing the 1px border from the notification panel, and using colour/contrast to bring attention to it. Maybe a ⚠️ icon or something like that too.

Glad you like it 🙂. The 1px border was my attempt to separate it from the surrounding controls and bring it to the front slightly. Happy to take another look if you have suggestions on other ways to achieve this using different colour/contrast combinations?

RussKie commented:
I was thinking about the icon too 🤔 

@NikolayXHD do you think you could help with getting colors right? So it would work with the dark theme as well.

NikolayXHD commented:
> I was thinking about the icon too 🤔
> 
> @NikolayXHD do you think you could help with getting colors right? So it would work with the dark theme as well.

If SystemColor is used, it's dark theme-compatible automatically. Non-system colors are ok too, because theming has methods to adapt arbitrary background / foreground colors to match the theme.

nick-bulleid commented:
On the icon, I could use the ⚠️ icon if there were merge conflicts and just an ℹ️ icon the rest of the time. Would you want to reuse the icons you already have or add a new ones?

RussKie commented:
> Would you want to reuse the icons you already have or add a new ones?

It is easier to reuse. But if you feel the existing icons aren't sufficient, feel free to add new icons.

drewnoakes commented:
> The 1px border was my attempt to separate it from the surrounding controls and bring it to the front slightly. Happy to take another look if you have suggestions on other ways to achieve this using different colour/contrast combinations?

Agree that visual separation is important, but consistency is important too and we don't use 1px borders any more.

Another observation is that the blue colour looks, at a glance, like a selected row.

Consider adding more padding to the box. That would further differentiate it from the rows below.

nick-bulleid commented:
I've updated the UI with
- Extra padding around the header
- A couple of icons that I reused from other parts of the UI
- Removed the 1px border

I've updated the original post with new images.

drewnoakes commented:
The screenshot looks fantastic.

pmiossec commented:
Great work!

It seems that you have to fix at least a unit test...

nick-bulleid commented:
> 
> 
> Great work!
> 
> It seems that you have to fix at least a unit test...

Done 👍

nick-bulleid commented on GitUI/GitUI.csproj:
```diff
@@ -617,6 +617,12 @@
     <Compile Include="UserControls\WaitSpinner.cs">
```
Done

RussKie commented:
I don't think it is working as intended.

I have a repo that is in a conflict state, but a version compiled from PR doesn't detect it.

Here's the 3.3.1 build:
![image](https://user-images.githubusercontent.com/4403806/72430156-b9a72980-37e5-11ea-9858-769dc934809e.png)

Here's a build from this PR:
![image](https://user-images.githubusercontent.com/4403806/72430200-d5123480-37e5-11ea-8f77-be2e02bc5b64.png)
Though if I open the commit dialog I get a notification (which isn't changed in this PR):
![image](https://user-images.githubusercontent.com/4403806/72430285-0c80e100-37e6-11ea-905c-d08d1188c588.png)

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,241 @@
+
+        private void ResolveButton_Click(object sender, EventArgs e)
+        {
+            GitModuleForm form = FindForm() as GitModuleForm;
```
I suggest to create a private property for this as it is used in few places.

nick-bulleid replied:
I've created a private property, but not sure if I'm using it correctly. Please advise if more changes are needed on this.

RussKie commented:
Here's the problem:
![image](https://user-images.githubusercontent.com/4403806/72431349-beb9a800-37e8-11ea-9b7a-b393309c5636.png)

nick-bulleid commented:
Interesting, I would have thought that the InMiddleOfMerge case would catch that.

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,265 @@
+
+        private void AbortButton_Click(object sender, EventArgs e)
+        {
+            if (!(Form is GitModuleForm))
```
```suggestion
            if (Form is null)
```
And other places too

RussKie commented on GitUI/UserControls/InteractiveGitActionControl.cs:
```diff
@@ -0,0 +1,265 @@
+            {
+                return FindForm() as GitModuleForm;
+            }
+        }
```
This can be shortened to one line: 
```cs
private GitModuleForm Form => FindForm() as GitModuleForm;
```

RussKie commented:
Looks better now, but we need to scale elements appropriately:
![image](https://user-images.githubusercontent.com/4403806/72719249-93113600-3bcb-11ea-97a3-2298a804bf08.png)

- DPI 144dpi (150% scaling)

pmiossec commented:
Did you already tried ` IconLabel.Size = DpiUtil.Scale( IconLabel.Size);`
at the end of the `InteractiveGitActionControl` constructor?

On Tue, 21 Jan 2020 at 10:05, Nick Bulleid <notifications@github.com> wrote:

> I'm having some trouble scaling the icon. Does anyone have any tips on how
> to achieve this?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7457?email_source=notifications&email_token=AADQLJEJERN4XTR6JADWMWLQ623E7A5CNFSM4JR6JG6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJO72TY#issuecomment-576585039>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQLJHXH2JQEZ62CMIHRMLQ623E7ANCNFSM4JR6JG6A>
> .
>

nickb-carallon commented:
> Did you already tried ` IconLabel.Size = DpiUtil.Scale( IconLabel.Size);` at the end of the `InteractiveGitActionControl` constructor?
> […](#)
> On Tue, 21 Jan 2020 at 10:05, Nick Bulleid ***@***.***> wrote: I'm having some trouble scaling the icon. Does anyone have any tips on how to achieve this? — You are receiving this because you commented. Reply to this email directly, view it on GitHub <#7457?email_source=notifications&email_token=AADQLJEJERN4XTR6JADWMWLQ623E7A5CNFSM4JR6JG6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJO72TY#issuecomment-576585039>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADQLJHXH2JQEZ62CMIHRMLQ623E7ANCNFSM4JR6JG6A> .

No, I was trying to use PerformAutoScale on the container. That sounds like a better solution though.

nick-bulleid commented:
No, I was trying to use PerformAutoScale on the container. That sounds like a better solution though.

> 
> Did you already tried ` IconLabel.Size = DpiUtil.Scale( IconLabel.Size);` at the end of the `InteractiveGitActionControl` constructor?
> […](#)
> On Tue, 21 Jan 2020 at 10:05, Nick Bulleid ***@***.***> wrote: I'm having some trouble scaling the icon. Does anyone have any tips on how to achieve this? — You are receiving this because you commented. Reply to this email directly, view it on GitHub <#7457?email_source=notifications&email_token=AADQLJEJERN4XTR6JADWMWLQ623E7A5CNFSM4JR6JG6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJO72TY#issuecomment-576585039>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADQLJHXH2JQEZ62CMIHRMLQ623E7ANCNFSM4JR6JG6A> .

nick-bulleid commented:
I found it worked if I switched to using a PictureBox. The Control should now display correctly at high dpi.

RussKie commented:
Looks good!
![image](https://user-images.githubusercontent.com/4403806/73259955-fc70f480-421c-11ea-9943-bb8c5ba4b9c4.png)

RussKie commented:
Thank you!
